### PR TITLE
Add support for new referencing resolution in jsonschema>=4.18

### LINKED ISSUE #3097: Code path in utils emits `jsonschema` deprecation warning
[jsonschema](https://pypi.org/project/jsonschema/) recently released a new version and introduced a new deprecation warning around `jsonschema.RefResolver` - this is utilized in `altair/utils/schemapi.py`.

You can replicate with the following code:
```python
import altair as alt

alt.StandardType("...")
```

```
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/vegalite/v4/schema/core.py:15771: in __init__
    super(StandardType, self).__init__(*args)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:197: in __init__
    self.to_dict(validate=True)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:358: in to_dict
    self.validate(result)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/altair/utils/schemapi.py:462: in validate
    resolver = jsonschema.RefResolver.from_schema(cls._rootschema or cls._schema)
/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/jsonschema/__init__.py:42: in __getattr__
    warnings.warn(
E   DeprecationWarning: jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the https://github.com/python-jsonschema/referencing library, which provides more compliant referencing behavior as well as more flexible APIs for customization. A future release will remove RefResolver. Please file a feature request (on referencing) if you are missing an API for the kind of customization you need.
```

---

Please follow these steps to make it more efficient to solve your issue:

- [x] Since Altair is a Python wrapper around the Vega-Lite visualization grammar, [most bugs should be reported directly to Vega-Lite](https://github.com/vega/vega-lite/issues). You can click the Action Button of your Altair chart and "Open in Vega Editor" to create a reproducible Vega-Lite example and see if you get the same error in the Vega Editor.
- [x] Search for duplicate issues.
- [x] Use the latest version of Altair.
- [x] Describe how to reproduce the bug and include the full code and data to reproduce it, ideally using a sample data set from `vega_datasets`.

### PR DESCRIPTION
**Pull request by binste**
Closes #3097.

@Julian:
* Could you elaborate on your comment in https://github.com/altair-viz/altair/issues/3097#issuecomment-1638584720? `alt.load_schema` should return the full Vega-Lite schema which is hosted at https://vega.github.io/schema/vega-lite/v5.8.0.json.:
```python
# It looks like you were previously picking a subschema of the "real" schema
# that lives at this URL -- I'll point out you don't really need to do that at
# this point, you should be able to take the "real" root schema as-is and use
# fragments to reference inside of it.
```
* In https://github.com/altair-viz/altair/pull/3118/files#diff-25906c2570b127f0bb9cff4550bcc153006b2fdb36aea55ae804ca9a74eb4404R176 I now iterate through a schema to append the URN I use in the registry to all `$ref`s. This works great with our test suite but it would be somehow great if there is a more direct way of telling that "all refs should always refer to a specific rootschema" without having to do this manually for all. Maybe this is a weird thing we have in Altair how we use the schemas. If you have any idea how to simplify this, it would be greatly appreciated. Else, we can also go forward with the current implementation.
* Do you spot anything in the diff which might not be a future-proof usage of `referencing`? I'm aware that the library is still in it's early stages so things might change but maybe you're already aware of an upcoming change.

Thanks again for your help! 

For myself:

- [x] Add changelog entry
- [x] Update test pipelines that they use the latest jsonschema version instead of hardcoded 4.17

### CODE REVIEW

File: altair/utils/schemapi.py
Comment (Julian) [inline altair/utils/schemapi.py:455]: I'll take a closer look at the whole PR tomorrow, but just because this is the first thing that pops out -- do you know why you do all this in the first place (pre-resolving references)? Doing so shouldn't be needed, and in fact is discouraged in the specification (basically because it's impossible in the general case) -- but it shouldn't be needed unless your schemas don't really have valid `$ref`s (in which case ideally you'd fix that in the schema, though maybe you don't control the Vega schemas?)

**Comment by Julian** [comment]
> Could you elaborate on your comment in https://github.com/altair-viz/altair/issues/3097#issuecomment-1638584720? alt.load_schema should return the full Vega-Lite schema which is hosted at https://vega.github.io/schema/vega-lite/v5.8.0.json.:

This was a mistake (or I looked at the wrong URL) -- all I meant was that it looked like `altair/vegalite/v5/schema/vega-lite-schema.json` had different contents than https://vega.github.io/schema/vega-lite/v5.8.0.json but I must have been mistaken, as `diff` says they're identical.

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:455]: I don't know why this happens here. I do see your point that this looks redundant at forst glance. So far I just tried to keep the functionality the same of this function. I won't get to this before mid of next week but then I can investigate what happens to our test suite and the code generation scripts if I remove this.

**Comment by mattijn** [comment]
A bit off-topic, but will the new dependency on `referencing` also include a hard dependency on https://github.com/crate-py/rpds? Hopefully I'm wrong, but as far as I can see this cannot be installed in a wasm-only environment.

**Comment by jheer** [comment]
> A bit off-topic, but will the new dependency on `referencing` also include a hard dependency on https://github.com/crate-py/rpds? Hopefully I'm wrong, but as far as I can see this cannot be installed in a wasm-only environment.

FWIW I’ve already run into this issue in the wild. As a result of this dependency Altair no longer loads successfully in Pyodide (WASM) via a direct micropip install. I have to explicitly install jsonschema in micropip first, pinned to an earlier release (< 4.18), to get Altair working again. If we enforce that Altair requires a later jsonschema version I don’t think WASM will work at all.

**Comment by Julian** [comment]
I haven't personally though I've seen that same page and had the same thought that it seems likely rpds would be buildable -- it's honestly not a priority for me myself to look into at this precise moment but I'm happy to take PRs if someone gets it working. And there's a strong chance in the future jsonschema will fallback to a pure Python implementation when that's faster (i.e. for PyPy) and if/when that's the case that too may be a solution.

**Comment by mattijn** [comment]
Since there is currently only pyodide support for `jsonschema<4.18` we should reflect this in our project.toml file (`"jsonschema>=3,<4.18"`).

**Comment by binste** [comment]
>  I'll defer to @binste, but I don't think we have any near-term plans to require jsonschema >=4.18 (we still support jsonschema 3 at the moment).

That's correct. For older versions of jsonschema, nothing should change. Altair will continue to support jsonschema >= 3.0 for the time being. I'd prefer to require >= 4.0 but we can't do that as DeepNote does not yet support it, see https://github.com/altair-viz/altair/issues/3038. This PR is only about supporting the new referencing library in case jsonschema >= 4.18 is installed.

jsonschema is already included in the [Pyodide distribution](https://pyodide.org/en/stable/usage/packages-in-pyodide.html) so I was hoping that rpds will be integrated as well once they build 4.18 but I don't know who decides that and how it works.

**Comment by jonmmease** [comment]
> Since there is currently only pyodide support for jsonschema<4.18 we should reflect this in our project.toml file ("jsonschema>=3,<4.18").

I don't think we want to put this constraint on Altair itself (at least after this PR is merged) just for the sake of Pyodide. As long as we're officially supporting and testing a version of jsonschema that's compatible with Pyodide, we can leave it to their community to decide how to handle their version constraints and upgrade path.

**Comment by mattijn** [comment]
I agree that avoiding an upper limit is beter. It did hurt a bit to write it down. 

It's good to hear that the pyodide community has some type of control on this.

(context, I tried to checkout this PR on an iPad using https://github.com/holzschu/Carnets but was running into the issues as mentioned above)

**Comment by jonmmease** [comment]
> but was running into the issues as mentioned above

Does it work if you install Altair with explicit jsonschema version pinned?

```
pip install altair==5.0.1 jsonschema==4.17
```

We could consider adding some documentation on Altair usage in pyodide, starting with the installation recommendations.  We could also consider a pyodide extras package (if micropip supports them). So that we could recommend something like

```
pip install "altair[pyodide]"
```

which would pin dependencies to known compatible versions, and perhaps install optional dependencies that are known to be compatible with pyodide.

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:222]: Thanks! Yeah that's what I originally did but not all schemas have a schema url attached so I had to revert to this. However, all schemas passed here are parts of the full Vega-Lite schema and use the same specification which is the one retrieved by `_get_json_schema_draft_url`.

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:455]: I looked into this and we need the recursive resolving not for the schema validation but for other things such as automatic code generation, dynamic population of methods on classes, etc. For example, class `alt.Y` needs to follow the json schema which can be found at `{'$ref': '#/definitions/PositionFieldDef'}` in the full Vega-Lite schema. We therefore only store that dictionary as reference on the Python class. However, in some parts of the code (e.g. https://github.com/altair-viz/altair/blob/main/altair/utils/schemapi.py#L1136), we need to know what is behind that references as we want to iterate through the schema in Python.

As part of a bigger rework, we could try to simplify this logic but that is outside of the scope of this PR.

**Comment by binste** [comment]
Thank you for all the inputs! Much appreciated. I like the idea of `pip install altair[pyodide]` and added it to #2818.

Regarding this PR, given that the tests pass including code generation, I think the current implementation is ready to be fully reviewed. But I'm still open to more elegant ways of handling the referencing resolution in case someone has a tip, see the bullet points in the PR description.

File: altair/utils/schemapi.py
Comment (mattijn) [reply altair/utils/schemapi.py:262]: How often is there no `$schema` defined when there should be? Is this something to be raised in the VL repo so we can include a reference as comment to simplify this later?

Comment (binste) [reply altair/utils/schemapi.py:262]: I think the VL schema is correct. `$schema` is defined once on the top-level which is enough as it is one JSON schema. In the code generation part in Altair, the generated Python classes have the `_schema` attribute set which refers to the corresponding definition in the VL schema. This attribute is set [here](https://github.com/altair-viz/altair/blob/main/tools/generate_schema_wrapper.py#L367) and you can see an example [here](https://github.com/altair-viz/altair/blob/main/altair/vegalite/v5/schema/core.py#L568). We don't set `$schema` in these generated dictionaries (`_schema`) which I think is fine as it would be the same across all of them. `_schema` is used in a few places such as during validation and for the property setter functionality, see [with_property_setters](https://github.com/altair-viz/altair/blob/main/altair/utils/schemapi.py#L1220).

Comment (mattijn) [reply altair/utils/schemapi.py:262]: Thanks for the clarification @binste! I remember you mentioned something about this in relation to the property setter before, but I could not find this comment anymore.

**Review by mattijn — APPROVED** [review]
I'm OK with the PR, but I only have read the code-diff a couple of times to understand what is going on. 
I did not actually test this PR. If this is important for you @binste, then it would be good to wait for a second approval of someone else.

**Comment by binste** [comment]
Thanks @mattijn for having a look! Apart from the test suite, I tested it also against a few more complex visualisations locally. I'll merge it now, maybe that increases the chance of someone finding edge cases before the release while we work on other PRs.

**Pull request merged**
